File: dsa in c/bubble_sort.c
#include <stdio.h>
#include <stdbool.h>
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}
void bubble_sort(int nums[], int n)
{
    for (int i = 0; i < n; i++)
    {
        bool is_swap = false;
        for (int j = 0; j < n - i - 1; j++)
        {
            if (nums[j] > nums[j + 1])
            {
                swap(&nums[j], &nums[j + 1]);
                is_swap = true;
            }
        }
        if (is_swap == false)
        {
            break;
        }
    }
}

int main()
{
    int nums[] = {4, 5, 6, 7, 1, 2};
    int n = sizeof(nums) / sizeof(nums[0]);
    bubble_sort(nums, n);
    for (int i = 0; i < n; i++)
    {
        printf("%d ", nums[i]);
    }
    return 0;
}

File: dsa in python/bubble_sort.py
# def bubble_sort(x):
#     n=len(x)
#     for i in range(n-2,-1,-1):
#         for j in range(0,i+1):
#             if x[j]>x[j+1]:
#                 x[j],x[j+1]=x[j+1],x[j]
# x=[7,6,5,4,3,2,1]
# bubble_sort(x)
# print(x) # tc = O(n^2)  sc=O(1)

def bubble_sort(x):
    n=len(x)
    for i in range(n-2,-1,-1):
        is_swap=False 
        for j in range(0,i+1):
            if x[j]>x[j+1]:
                x[j],x[j+1]=x[j+1],x[j]
                is_swap=True
        if is_swap==False:
            break
        print(x)
x=[5,8,1,6,9,2,4]
bubble_sort(x)
print(x) # Best Case
